### ISSUE TITLE
Independent Categorical Sampling

### ISSUE THREAD

**Issue by stergiosba** [google/jax#18950]
### Description

Hello guys, thank you so much for creating JAX it is really a marvelous tool.

Sorry if this is basic but I face the following confusing issue.

I have two Categorical random variables X, Y that are independent. What is the correct way to sample the two distributions?
I provide a short program that shows what I mean. 

Say we do the following experiments E1 and E2:
```python
import jax.numpy as jnp
import jax.random as jrandom
import matplotlib.pyplot as plt

n = 10
key_X = jrandom.PRNGKey(120)
key_Y = jrandom.PRNGKey(845)
logits_X = jrandom.uniform(key_X, shape=(n,))
logits_Y = jrandom.uniform(key_Y, shape=(n,))

joint_logits = jnp.array([logits_X, logits_Y])

E1 = []
E2 = []

key = jrandom.PRNGKey(0)
n_s = 50
for _ in range(n_s):
    e1 = [jrandom.categorical(key, logits) for logits in joint_logits]
    e2 = jrandom.categorical(key, joint_logits)
    
    E1.append(e1)
    E2.append(e2)
    key, subkey = jrandom.split(key)

E1 = jnp.array(E1)
E2 = jnp.array(E2)

plt.hist(E2[:,0], bins=10*n)
plt.hist(E1[:,0], bins=10*n)
plt.grid()
plt.title("Samples of X")
plt.xlabel("Sample value of X")
plt.ylabel("Frequency")
plt.legend(["Experiment 1 (loop)", "Experiment 2 (stacked)"])
plt.show()

plt.hist(E1[:,1], bins=10*n)
plt.hist(E2[:,1], bins=10*n)
plt.grid()
plt.title("Samples of Y")
plt.xlabel("Sample value of Y")
plt.ylabel("Frequency")
plt.legend(["Experiment 1 (loop)", "Experiment 2 (stacked)"])
plt.show()
``````
![image](https://github.com/google/jax/assets/29761549/717894c1-5602-4c92-8d17-324d71d0ca02)
![image](https://github.com/google/jax/assets/29761549/10add5ca-8242-4bda-8b0f-f08f57b86713)

I looked at the code and I believe that the whole thing boils down to sampling a uniform distribution in two different ways:
1) Two times with the same key resulting in Experiment E1
2) Two times with a different key but these two keys are somehow connected resulting in Experiment E2.

Based on what the docs say in [Pseudo Random Numbers in JAX](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html) I understand that I need to manually split the key before sampling. Is the the process followed in Experiment 2 essentially doing the same thing under the hood and therefore is the right way to go?

I do realize that E1 is almost certainly wrong as the same key is used and therefore the two components in `E1` are dependent.

I appreciate any help.
Thank you!

### What jax/jaxlib version are you using?

jax 4.21, jaxlib 4.21, (0.4.21 0.4.21)

### Which accelerator(s) are you using?

GPU

### Additional system info?

Python : 3.10

### NVIDIA GPU info

NVIDIA GeForce RTX 3070 Ti
Driver Version: 535.129.03 
CUDA Version: 12.2

**Comment by stergiosba** [comment]
This is not an issue, it was intended behavior.

You can go ahead and delete it even if you think it will confuse people.

**Comment by jakevdp** [comment]
Hi - just a note on your code: your key splitting as written is not actually preventing reuse of keys: you're passing the same key to `categorical` multiple times within each iteration I would write the for loop something like this:
```python
for _ in range(n_s):
    key, subkey, *subkeys = jrandom.split(key, 2 + len(joint_logits)]
    e1 = [jrandom.categorical(k, logits) for k, logits in zip(subkeys, joint_logits)]
    e2 = jrandom.categorical(subkey, joint_logits)
    
    E1.append(e1)
    E2.append(e2)
```
This way, each call to a `jax.random` API gets a fresh unused key, and you won't end up with unintended correlations between your different random draws.

**Comment by stergiosba** [comment]
Hey,

Thank you very much for the explanation.
